### PR TITLE
Add End-to-End CI Workflow for Treadmill

### PR DESCRIPTION
**Pull request by charles37**
### Summary:
This PR introduces a comprehensive end-to-end (E2E) CI workflow for the Treadmill project. The new workflow, named "treadmill-end-to-end-ci", automates the testing process and ensures the reliability of our API and core functionalities.
# Key features of this CI workflow:
- Sets up a PostgreSQL database service for testing.
- Builds all necessary components (switchboard, mock supervisor, and puppet).
- Generates and configures test environments.
- Performs E2E API tests, including job creation, status checking, and cancellation.
  - The E2E tests the API's ability to create a job, check its status, and cancel it, while verifying proper authentication throughout. The tests ensure that job creation returns a valid ID, that status checks reflect the job's current state, and that cancellation results in a "finished" status.

# Next steps:
- Implement the TODO sections for database interaction tests and puppet communication tests.
- Consider expanding the test coverage to include more edge cases and error scenarios.
- Regularly review and update the CI workflow as the project evolves.

### CODE REVIEW

File: shell.nix
Comment (lschuermann) [inline shell.nix:1]: Should this file be added as part of this PR?

File: supervisor/mock/config.example.toml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [base]
 supervisor_id = "7d55ec6d-15e7-4b84-8c04-7c085fe60df4"
-coord_connector = "cli_connector"
+coord_connector = "ws_connector"
```
Comment (lschuermann) [inline supervisor/mock/config.example.toml:3]: I think leaving this set to the CLI connector by default is a good idea -- it makes it easier to test local changes. We should be able to override this parameter in the CI workflow?

**Review by lschuermann — COMMENTED** [review]
This is really cool, and the workflow is surprisingly readable!

Would it be a good idea to split out the end-to-end test in its own `.yml` workflow file? I'm not sure whether that buys us much, but it might be nice to split tests in static checks, and actual integration tests. It also makes the base `ci.yml` file for static checks more readable.

File: shell.nix
Comment (charles37) [reply shell.nix:1]: I thought it would go along with the idea of having devops/ci stuff, but I can make it a separate PR.

Comment (lschuermann) [reply shell.nix:1]: Separate PR would be great!

File: .github/workflows/ci.yml
Comment (lschuermann) [inline .github/workflows/ci.yml:1]: Can you just revert all changes to this file?

Something like `git checkout main -- .github/workflows/ci.yml`?

**Review by lschuermann — DISMISSED** [review]
Looks good otherwise.
